File: backend/functions/src/cleanup_history.test.ts
import * as admin from "firebase-admin";
import { processHistoryCleanup } from "./cleanup_history";

describe("processHistoryCleanup", () => {
  it("deletes expired history documents and commits in batches", async () => {
    const mockRef1 = { path: "users/u1/history/d1" };
    const mockRef2 = { path: "users/u2/history/d2" };
    const mockDocs = [
      { id: "d1", ref: mockRef1 },
      { id: "d2", ref: mockRef2 },
    ];

    const mockDelete = jest.fn();
    const mockCommit = jest.fn().mockResolvedValue([]);
    const mockBatch = jest.fn(() => ({
      delete: mockDelete,
      commit: mockCommit,
    }));

    const mockGet = jest.fn().mockResolvedValue({
      empty: false,
      size: 2,
      docs: mockDocs,
    });

    const mockQuery = {
      where: jest.fn().mockReturnThis(),
      limit: jest.fn().mockReturnThis(),
      get: mockGet,
    };

    const mockDb = {
      collectionGroup: jest.fn().mockReturnValue(mockQuery),
      batch: mockBatch,
    } as unknown as admin.firestore.Firestore;

    const testNow = admin.firestore.Timestamp.fromDate(new Date("2026-08-31T00:00:00Z"));
    const result = await processHistoryCleanup(mockDb, testNow, 500);

    expect(mockDb.collectionGroup).toHaveBeenCalledWith("history");
    expect(mockQuery.where).toHaveBeenCalledWith("expiresAt", "<=", testNow);
    expect(mockQuery.limit).toHaveBeenCalledWith(500);
    expect(mockDelete).toHaveBeenCalledWith(mockRef1);
    expect(mockDelete).toHaveBeenCalledWith(mockRef2);
    expect(mockCommit).toHaveBeenCalledTimes(1);
    expect(result.success).toBe(true);
    expect(result.totalDeleted).toBe(2);
    expect(result.batchesProcessed).toBe(1);
    expect(result.durationMs).toBeGreaterThanOrEqual(0);
  });

  it("handles pagination / multi-batch deletion when expired documents exceed single batch limit", async () => {
    const firstBatchDocs = Array.from({ length: 500 }, (_, i) => ({
      id: `doc_b1_${i}`,
      ref: { path: `users/u1/history/doc_b1_${i}` },
    }));
    const secondBatchDocs = Array.from({ length: 250 }, (_, i) => ({
      id: `doc_b2_${i}`,
      ref: { path: `users/u2/history/doc_b2_${i}` },
    }));

    const mockDelete = jest.fn();
    const mockCommit = jest.fn().mockResolvedValue([]);
    const mockBatch = jest.fn(() => ({
      delete: mockDelete,
      commit: mockCommit,
    }));

    const mockGet = jest
      .fn()
      .mockResolvedValueOnce({
        empty: false,
        size: 500,
        docs: firstBatchDocs,
      })
      .mockResolvedValueOnce({
        empty: false,
        size: 250,
        docs: secondBatchDocs,
      });

    const mockQuery = {
      where: jest.fn().mockReturnThis(),
      limit: jest.fn().mockReturnThis(),
      get: mockGet,
    };

    const mockDb = {
      collectionGroup: jest.fn().mockReturnValue(mockQuery),
      batch: mockBatch,
    } as unknown as admin.firestore.Firestore;

    const testNow = admin.firestore.Timestamp.now();
    const result = await processHistoryCleanup(mockDb, testNow, 500);

    expect(mockGet).toHaveBeenCalledTimes(2);
    expect(mockBatch).toHaveBeenCalledTimes(2);
    expect(mockDelete).toHaveBeenCalledTimes(750);
    expect(mockCommit).toHaveBeenCalledTimes(2);
    expect(result.success).toBe(true);
    expect(result.totalDeleted).toBe(750);
    expect(result.batchesProcessed).toBe(2);
  });

  it("handles no-op scenario when no documents have expired", async () => {
    const mockDelete = jest.fn();
    const mockCommit = jest.fn();
    const mockBatch = jest.fn(() => ({
      delete: mockDelete,
      commit: mockCommit,
    }));

    const mockGet = jest.fn().mockResolvedValue({
      empty: true,
      size: 0,
      docs: [],
    });

    const mockQuery = {
      where: jest.fn().mockReturnThis(),
      limit: jest.fn().mockReturnThis(),
      get: mockGet,
    };

    const mockDb = {
      collectionGroup: jest.fn().mockReturnValue(mockQuery),
      batch: mockBatch,
    } as unknown as admin.firestore.Firestore;

    const testNow = admin.firestore.Timestamp.now();
    const result = await processHistoryCleanup(mockDb, testNow, 500);

    expect(mockGet).toHaveBeenCalledTimes(1);
    expect(mockBatch).not.toHaveBeenCalled();
    expect(mockDelete).not.toHaveBeenCalled();
    expect(mockCommit).not.toHaveBeenCalled();
    expect(result.success).toBe(true);
    expect(result.totalDeleted).toBe(0);
    expect(result.batchesProcessed).toBe(0);
  });

  it("respects maxBatches limit to prevent unbounded loops", async () => {
    const fullBatchDocs = Array.from({ length: 100 }, (_, i) => ({
      id: `doc_${i}`,
      ref: { path: `users/u1/history/doc_${i}` },
    }));

    const mockDelete = jest.fn();
    const mockCommit = jest.fn().mockResolvedValue([]);
    const mockBatch = jest.fn(() => ({
      delete: mockDelete,
      commit: mockCommit,
    }));

    const mockGet = jest.fn().mockResolvedValue({
      empty: false,
      size: 100,
      docs: fullBatchDocs,
    });

    const mockQuery = {
      where: jest.fn().mockReturnThis(),
      limit: jest.fn().mockReturnThis(),
      get: mockGet,
    };

    const mockDb = {
      collectionGroup: jest.fn().mockReturnValue(mockQuery),
      batch: mockBatch,
    } as unknown as admin.firestore.Firestore;

    const testNow = admin.firestore.Timestamp.now();
    // Set maxBatches to 3
    const result = await processHistoryCleanup(mockDb, testNow, 100, 3);

    expect(mockGet).toHaveBeenCalledTimes(3);
    expect(mockCommit).toHaveBeenCalledTimes(3);
    expect(result.success).toBe(true);
    expect(result.totalDeleted).toBe(300);
    expect(result.batchesProcessed).toBe(3);
  });

  it("handles and logs error when batch commit fails", async () => {
    const mockRef = { path: "users/u1/history/d1" };
    const mockDocs = [{ id: "d1", ref: mockRef }];

    const commitError = new Error("Firestore batch commit failed due to connection error");
    const mockDelete = jest.fn();
    const mockCommit = jest.fn().mockRejectedValue(commitError);
    const mockBatch = jest.fn(() => ({
      delete: mockDelete,
      commit: mockCommit,
    }));

    const mockGet = jest.fn().mockResolvedValue({
      empty: false,
      size: 1,
      docs: mockDocs,
    });

    const mockQuery = {
      where: jest.fn().mockReturnThis(),
      limit: jest.fn().mockReturnThis(),
      get: mockGet,
    };

    const mockDb = {
      collectionGroup: jest.fn().mockReturnValue(mockQuery),
      batch: mockBatch,
    } as unknown as admin.firestore.Firestore;

    const testNow = admin.firestore.Timestamp.now();
    await expect(processHistoryCleanup(mockDb, testNow, 500)).rejects.toThrow(
      "Firestore batch commit failed due to connection error"
    );
  });
});

File: backend/functions/test/cleanup_history_test.dart
import 'package:functions/src/handlers/cleanup_history.dart';
import 'package:test/test.dart';
import 'test_helpers.dart';

void main() {
  group('processHistoryCleanup', () {
    test('deletes expired history documents and commits in batches', () async {
      final mockDb = MockFirestoreDatabase();
      final ref1 = MockDocumentReference('d1', path: 'users/u1/history/d1');
      final ref2 = MockDocumentReference('d2', path: 'users/u2/history/d2');
      final mockDocs = [
        MockDocumentSnapshot('d1', {}, true, ref1),
        MockDocumentSnapshot('d2', {}, true, ref2),
      ];

      final mockQuery = MockQuery();
      mockQuery.cannedDocs = mockDocs;

      mockDb.onCollectionGroup = (name) {
        expect(name, equals('history'));
        return mockQuery;
      };

      final testNow = DateTime.parse('2026-08-31T00:00:00Z');
      final result = await processHistoryCleanup(mockDb, testNow, 500);

      expect(result.success, isTrue);
      expect(result.totalDeleted, equals(2));
      expect(result.batchesProcessed, equals(1));
      expect(result.durationMs, greaterThanOrEqualTo(0));
      expect(mockDb.committedBatches.length, equals(1));
      expect(
          mockDb.committedBatches.first.deletedRefs, containsAll([ref1, ref2]));
    });

    test(
        'handles pagination / multi-batch deletion when expired documents exceed single batch limit',
        () async {
      final mockDb = MockFirestoreDatabase();
      final firstBatchDocs = List.generate(
        500,
        (i) => MockDocumentSnapshot(
          'doc_b1_$i',
          {},
          true,
          MockDocumentReference('doc_b1_$i',
              path: 'users/u1/history/doc_b1_$i'),
        ),
      );
      final secondBatchDocs = List.generate(
        250,
        (i) => MockDocumentSnapshot(
          'doc_b2_$i',
          {},
          true,
          MockDocumentReference('doc_b2_$i',
              path: 'users/u2/history/doc_b2_$i'),
        ),
      );

      var callCount = 0;
      final mockQuery = MockQuery();
      mockQuery.onGet = () async {
        callCount++;
        if (callCount == 1) {
          return MockQuerySnapshot(firstBatchDocs);
        } else if (callCount == 2) {
          return MockQuerySnapshot(secondBatchDocs);
        }
        return MockQuerySnapshot([]);
      };

      mockDb.onCollectionGroup = (name) => mockQuery;

      final testNow = DateTime.now().toUtc();
      final result = await processHistoryCleanup(mockDb, testNow, 500);

      expect(callCount, equals(2));
      expect(result.success, isTrue);
      expect(result.totalDeleted, equals(750));
      expect(result.batchesProcessed, equals(2));
      expect(mockDb.committedBatches.length, equals(2));
    });

    test('handles no-op scenario when no documents have expired', () async {
      final mockDb = MockFirestoreDatabase();
      final mockQuery = MockQuery();
      mockQuery.cannedDocs = [];
      mockDb.onCollectionGroup = (name) => mockQuery;

      final testNow = DateTime.now().toUtc();
      final result = await processHistoryCleanup(mockDb, testNow, 500);

      expect(result.success, isTrue);
      expect(result.totalDeleted, equals(0));
      expect(result.batchesProcessed, equals(0));
      expect(mockDb.committedBatches, isEmpty);
    });

    test('respects maxBatches limit to prevent unbounded loops', () async {
      final mockDb = MockFirestoreDatabase();
      final fullBatchDocs = List.generate(
        100,
        (i) => MockDocumentSnapshot(
          'doc_$i',
          {},
          true,
          MockDocumentReference('doc_$i', path: 'users/u1/history/doc_$i'),
        ),
      );

      var callCount = 0;
      final mockQuery = MockQuery();
      mockQuery.onGet = () async {
        callCount++;
        return MockQuerySnapshot(fullBatchDocs);
      };

      mockDb.onCollectionGroup = (name) => mockQuery;

      final testNow = DateTime.now().toUtc();
      final result = await processHistoryCleanup(mockDb, testNow, 100, 3);

      expect(callCount, equals(3));
      expect(result.success, isTrue);
      expect(result.totalDeleted, equals(300));
      expect(result.batchesProcessed, equals(3));
      expect(mockDb.committedBatches.length, equals(3));
    });

    test('handles and logs error when batch commit fails', () async {
      final mockDb = MockFirestoreDatabase();
      final mockDocs = [
        MockDocumentSnapshot(
          'd1',
          {},
          true,
          MockDocumentReference('d1', path: 'users/u1/history/d1'),
        ),
      ];

      final mockQuery = MockQuery();
      mockQuery.cannedDocs = mockDocs;
      mockDb.onCollectionGroup = (name) => mockQuery;

      // Make batch commit fail
      final batch = MockWriteBatch(onCommitAsync: () async {
        throw Exception(
            'Firestore batch commit failed due to connection error');
      });
      mockDb.onBatch = () => batch;

      final testNow = DateTime.now().toUtc();
      expect(
        () async => await processHistoryCleanup(mockDb, testNow, 500),
        throwsA(isA<Exception>()),
      );
    });
  });
}
